In [24]:
import pandas as pd
import numpy as np

### I. Working with Datetime

In [25]:
df = pd.read_csv(r"src\hourlySteps_merged.csv")
df.dtypes

Id              int64
ActivityHour      str
StepTotal       int64
dtype: object

In [26]:
df.head(10)

,Id,ActivityHour,StepTotal
0,1503960366,3/12/2016 12:00:00 AM,0
1,1503960366,3/12/2016 1:00:00 AM,0
2,1503960366,3/12/2016 2:00:00 AM,0
3,1503960366,3/12/2016 3:00:00 AM,0
4,1503960366,3/12/2016 4:00:00 AM,0
5,1503960366,3/12/2016 5:00:00 AM,0
6,1503960366,3/12/2016 6:00:00 AM,0
7,1503960366,3/12/2016 7:00:00 AM,0
8,1503960366,3/12/2016 8:00:00 AM,0
9,1503960366,3/12/2016 9:00:00 AM,8


In [27]:
#Convert the string representation of the date/time into a proper pandas datetime object using pd.to_datetime().
df['ActivityHour_datetime'] = pd.to_datetime(df['ActivityHour'], format='%m/%d/%Y %I:%M:%S %p')
df.head(10)

,Id,ActivityHour,StepTotal,ActivityHour_datetime
0,1503960366,3/12/2016 12:00:00 AM,0,2016-03-12 00:00:00
1,1503960366,3/12/2016 1:00:00 AM,0,2016-03-12 01:00:00
2,1503960366,3/12/2016 2:00:00 AM,0,2016-03-12 02:00:00
3,1503960366,3/12/2016 3:00:00 AM,0,2016-03-12 03:00:00
4,1503960366,3/12/2016 4:00:00 AM,0,2016-03-12 04:00:00
5,1503960366,3/12/2016 5:00:00 AM,0,2016-03-12 05:00:00
6,1503960366,3/12/2016 6:00:00 AM,0,2016-03-12 06:00:00
7,1503960366,3/12/2016 7:00:00 AM,0,2016-03-12 07:00:00
8,1503960366,3/12/2016 8:00:00 AM,0,2016-03-12 08:00:00
9,1503960366,3/12/2016 9:00:00 AM,8,2016-03-12 09:00:00


In [28]:
df.dtypes

Id                                int64
ActivityHour                        str
StepTotal                         int64
ActivityHour_datetime    datetime64[us]
dtype: object

In [29]:
#Extract at least three new useful features from the datetime column (e.g., year, month, day of the week, hour, or an is_weekend boolean flag).
df["Hour"] = df["ActivityHour_datetime"].dt.hour
df["Day"] = df["ActivityHour_datetime"].dt.day
df["Month"] = df["ActivityHour_datetime"].dt.month
df["Year"] = df["ActivityHour_datetime"].dt.year

df["Day_of_Week"] = df["ActivityHour_datetime"].dt.day_name()
df["Is_Weekend"] = df["ActivityHour_datetime"].dt.weekday > 4

df.head(10)

,Id,ActivityHour,StepTotal,ActivityHour_datetime,Hour,Day,Month,Year,Day_of_Week,Is_Weekend
0,1503960366,3/12/2016 12:00:00 AM,0,2016-03-12 00:00:00,0,12,3,2016,Saturday,True
1,1503960366,3/12/2016 1:00:00 AM,0,2016-03-12 01:00:00,1,12,3,2016,Saturday,True
2,1503960366,3/12/2016 2:00:00 AM,0,2016-03-12 02:00:00,2,12,3,2016,Saturday,True
3,1503960366,3/12/2016 3:00:00 AM,0,2016-03-12 03:00:00,3,12,3,2016,Saturday,True
4,1503960366,3/12/2016 4:00:00 AM,0,2016-03-12 04:00:00,4,12,3,2016,Saturday,True
5,1503960366,3/12/2016 5:00:00 AM,0,2016-03-12 05:00:00,5,12,3,2016,Saturday,True
6,1503960366,3/12/2016 6:00:00 AM,0,2016-03-12 06:00:00,6,12,3,2016,Saturday,True
7,1503960366,3/12/2016 7:00:00 AM,0,2016-03-12 07:00:00,7,12,3,2016,Saturday,True
8,1503960366,3/12/2016 8:00:00 AM,0,2016-03-12 08:00:00,8,12,3,2016,Saturday,True
9,1503960366,3/12/2016 9:00:00 AM,8,2016-03-12 09:00:00,9,12,3,2016,Saturday,True


In [30]:
df.dtypes

Id                                int64
ActivityHour                        str
StepTotal                         int64
ActivityHour_datetime    datetime64[us]
Hour                              int32
Day                               int32
Month                             int32
Year                              int32
Day_of_Week                         str
Is_Weekend                         bool
dtype: object

In [31]:
#Drop the original raw datetime string column from the working DataFrame.
df = df.drop(columns = ['ActivityHour'])
df.head(10)

,Id,StepTotal,ActivityHour_datetime,Hour,Day,Month,Year,Day_of_Week,Is_Weekend
0,1503960366,0,2016-03-12 00:00:00,0,12,3,2016,Saturday,True
1,1503960366,0,2016-03-12 01:00:00,1,12,3,2016,Saturday,True
2,1503960366,0,2016-03-12 02:00:00,2,12,3,2016,Saturday,True
3,1503960366,0,2016-03-12 03:00:00,3,12,3,2016,Saturday,True
4,1503960366,0,2016-03-12 04:00:00,4,12,3,2016,Saturday,True
5,1503960366,0,2016-03-12 05:00:00,5,12,3,2016,Saturday,True
6,1503960366,0,2016-03-12 06:00:00,6,12,3,2016,Saturday,True
7,1503960366,0,2016-03-12 07:00:00,7,12,3,2016,Saturday,True
8,1503960366,0,2016-03-12 08:00:00,8,12,3,2016,Saturday,True
9,1503960366,8,2016-03-12 09:00:00,9,12,3,2016,Saturday,True


In [32]:
df.dtypes

Id                                int64
StepTotal                         int64
ActivityHour_datetime    datetime64[us]
Hour                              int32
Day                               int32
Month                             int32
Year                              int32
Day_of_Week                         str
Is_Weekend                         bool
dtype: object

#### Second DF to work with timestamps

In [33]:
df = pd.read_csv(r"src\synthetic_food_delivery_transactions.csv")
df.dtypes

order_id                          int64
customer_id                       int64
customer_loyalty_tier               str
customer_acquisition_channel        str
customer_signup_date                str
restaurant_id                     int64
restaurant_cuisine                  str
restaurant_city                     str
restaurant_rating               float64
restaurant_avg_prep_time          int64
menu_item_id                      int64
menu_item_name                      str
menu_item_category                  str
item_price                      float64
quantity                          int64
order_timestamp                     str
delivery_timestamp                  str
delivery_duration_actual        float64
delivery_duration_estimated       int64
delivery_fee                    float64
tip                             float64
total_amount                    float64
discount_code                       str
discount_type                       str
discount_value                  float64


In [34]:
df.head(10)

,order_id,customer_id,customer_loyalty_tier,customer_acquisition_channel,customer_signup_date,restaurant_id,restaurant_cuisine,restaurant_city,restaurant_rating,restaurant_avg_prep_time,...,total_amount,discount_code,discount_type,discount_value,order_status,weather_temperature,weather_precipitation,is_weekend,is_holiday,event_flag
0,4002,925,Gold,social_media,2024-07-22,124,American,Los Angeles,3.6,23,...,25.42,PROMO-BFUZ,percentage,4.97,completed,2.9,25.6,True,True,NaN
1,4879,3163,Bronze,referral,2023-03-17,43,Chinese,Houston,4.8,32,...,22.70,NaN,NaN,0.00,completed,4.0,0.4,True,True,NaN
2,1124,4154,Bronze,referral,2023-01-14,188,American,San Antonio,3.4,11,...,27.74,NaN,NaN,0.00,completed,4.4,0.0,True,True,NaN
3,6764,4530,Bronze,email,2024-02-26,91,Mediterranean,New York,3.1,17,...,31.85,NaN,NaN,0.00,completed,9.9,0.0,True,True,NaN
4,9073,2570,Bronze,paid_ads,2024-08-25,84,Mediterranean,San Antonio,3.2,28,...,28.46,NaN,NaN,0.00,completed,7.0,16.1,True,True,NaN
5,7005,2816,Bronze,email,2023-05-28,161,American,Houston,3.8,15,...,64.00,NaN,NaN,0.00,completed,2.3,1.5,True,True,NaN
6,9422,6973,Bronze,paid_ads,2024-06-24,53,Fast Food,Houston,4.7,15,...,139.07,NaN,NaN,0.00,completed,6.8,0.5,True,True,NaN
7,7068,4579,Gold,paid_ads,2023-11-29,116,Indian,San Diego,4.4,12,...,37.19,NaN,NaN,0.00,completed,1.2,0.0,True,True,NaN
8,9274,4463,Silver,referral,2024-07-26,187,Chinese,Chicago,3.6,10,...,38.93,PROMO-XWIT,percentage,5.14,completed,1.6,0.0,True,True,NaN
9,1942,889,Silver,paid_ads,2023-12-29,64,Italian,New York,4.0,15,...,18.76,NaN,NaN,0.00,completed,2.9,0.0,True,True,NaN


In [35]:
timestamp_df = pd.DataFrame()

timestamp_df['Order_ID'] = df['order_id']
timestamp_df['OrderTimestamp'] = pd.to_datetime(df['order_timestamp'])
timestamp_df['DeliveryTimestamp'] = pd.to_datetime(df['delivery_timestamp'])

timestamp_df['DeliveryTimestamp'] = pd.to_datetime(df['delivery_timestamp'])
timestamp_df['OrderTimestamp'] = pd.to_datetime(df['order_timestamp'])

timestamp_df.head(10)

,Order_ID,OrderTimestamp,DeliveryTimestamp
0,4002,2023-01-01 00:10:17,2023-01-01 01:09:17
1,4879,2023-01-01 00:29:15,2023-01-01 01:26:15
2,1124,2023-01-01 04:06:27,2023-01-01 04:47:27
3,6764,2023-01-01 05:05:28,2023-01-01 05:55:28
4,9073,2023-01-01 08:46:51,2023-01-01 09:55:51
5,7005,2023-01-01 10:33:29,2023-01-01 11:10:29
6,9422,2023-01-01 11:04:47,2023-01-01 11:48:47
7,7068,2023-01-01 15:48:18,2023-01-01 16:31:18
8,9274,2023-01-01 17:51:54,2023-01-01 18:32:54
9,1942,2023-01-01 18:16:23,2023-01-01 18:43:23


In [36]:
timestamp_df["TotalDeliveryTime"] = timestamp_df["DeliveryTimestamp"] - timestamp_df["OrderTimestamp"]
timestamp_df.head(10)

,Order_ID,OrderTimestamp,DeliveryTimestamp,TotalDeliveryTime
0,4002,2023-01-01 00:10:17,2023-01-01 01:09:17,0 days 00:59:00
1,4879,2023-01-01 00:29:15,2023-01-01 01:26:15,0 days 00:57:00
2,1124,2023-01-01 04:06:27,2023-01-01 04:47:27,0 days 00:41:00
3,6764,2023-01-01 05:05:28,2023-01-01 05:55:28,0 days 00:50:00
4,9073,2023-01-01 08:46:51,2023-01-01 09:55:51,0 days 01:09:00
5,7005,2023-01-01 10:33:29,2023-01-01 11:10:29,0 days 00:37:00
6,9422,2023-01-01 11:04:47,2023-01-01 11:48:47,0 days 00:44:00
7,7068,2023-01-01 15:48:18,2023-01-01 16:31:18,0 days 00:43:00
8,9274,2023-01-01 17:51:54,2023-01-01 18:32:54,0 days 00:41:00
9,1942,2023-01-01 18:16:23,2023-01-01 18:43:23,0 days 00:27:00


In [37]:
timestamp_df.dtypes

Order_ID                       int64
OrderTimestamp        datetime64[us]
DeliveryTimestamp     datetime64[us]
TotalDeliveryTime    timedelta64[us]
dtype: object

### II. Working with categorical data

In [38]:
categorical_df = df[['customer_loyalty_tier', 'restaurant_city', 'restaurant_cuisine']]
categorical_df.head(10)

,customer_loyalty_tier,restaurant_city,restaurant_cuisine
0,Gold,Los Angeles,American
1,Bronze,Houston,Chinese
2,Bronze,San Antonio,American
3,Bronze,New York,Mediterranean
4,Bronze,San Antonio,Mediterranean
5,Bronze,Houston,American
6,Bronze,Houston,Fast Food
7,Gold,San Diego,Indian
8,Silver,Chicago,Chinese
9,Silver,New York,Italian


In [39]:
#Identify nominal (unordered) and ordinal (ordered) categorical columns in your dataset.
#Ordered column
categorical_df["customer_loyalty_tier"].unique()

<StringArray>
['Gold', 'Bronze', 'Silver']
Length: 3, dtype: str

In [40]:
#Unordered columns
categorical_df["restaurant_city"].unique()

<StringArray>
[ 'Los Angeles',      'Houston',  'San Antonio',     'New York',
    'San Diego',      'Chicago',      'Phoenix', 'Philadelphia']
Length: 8, dtype: str

In [41]:
categorical_df["restaurant_cuisine"].unique()

<StringArray>
[     'American',       'Chinese', 'Mediterranean',     'Fast Food',
        'Indian',       'Italian',      'Japanese',       'Mexican']
Length: 8, dtype: str

In [42]:
#Apply Label Encoding (or manual mapping) to at least one ordinal categorical column (e.g., mapping 'low', 'medium', 'high' to 0, 1, 2).
loyalty_map = {'Bronze': 0, 'Silver': 1, 'Gold': 2}
mapped_df = categorical_df
mapped_df['Mapped_Loyalty'] = categorical_df['customer_loyalty_tier'].map(loyalty_map)
mapped_df.head(10)

,customer_loyalty_tier,restaurant_city,restaurant_cuisine,Mapped_Loyalty
0,Gold,Los Angeles,American,2
1,Bronze,Houston,Chinese,0
2,Bronze,San Antonio,American,0
3,Bronze,New York,Mediterranean,0
4,Bronze,San Antonio,Mediterranean,0
5,Bronze,Houston,American,0
6,Bronze,Houston,Fast Food,0
7,Gold,San Diego,Indian,2
8,Silver,Chicago,Chinese,1
9,Silver,New York,Italian,1


In [43]:
mapped_df.dtypes

customer_loyalty_tier      str
restaurant_city            str
restaurant_cuisine         str
Mapped_Loyalty           int64
dtype: object

In [44]:
#Apply One-Hot Encoding (using pd.get_dummies() or sklearn's OneHotEncoder) to at least one nominal categorical column.
encoded_df = pd.get_dummies(categorical_df, columns=['restaurant_city', 'restaurant_cuisine'])
encoded_df.head(10)

,customer_loyalty_tier,Mapped_Loyalty,restaurant_city_Chicago,restaurant_city_Houston,restaurant_city_Los Angeles,restaurant_city_New York,restaurant_city_Philadelphia,restaurant_city_Phoenix,restaurant_city_San Antonio,restaurant_city_San Diego,restaurant_cuisine_American,restaurant_cuisine_Chinese,restaurant_cuisine_Fast Food,restaurant_cuisine_Indian,restaurant_cuisine_Italian,restaurant_cuisine_Japanese,restaurant_cuisine_Mediterranean,restaurant_cuisine_Mexican
0,Gold,2,False,False,True,False,False,False,False,False,True,False,False,False,False,False,False,False
1,Bronze,0,False,True,False,False,False,False,False,False,False,True,False,False,False,False,False,False
2,Bronze,0,False,False,False,False,False,False,True,False,True,False,False,False,False,False,False,False
3,Bronze,0,False,False,False,True,False,False,False,False,False,False,False,False,False,False,True,False
4,Bronze,0,False,False,False,False,False,False,True,False,False,False,False,False,False,False,True,False
5,Bronze,0,False,True,False,False,False,False,False,False,True,False,False,False,False,False,False,False
6,Bronze,0,False,True,False,False,False,False,False,False,False,False,True,False,False,False,False,False
7,Gold,2,False,False,False,False,False,False,False,True,False,False,False,True,False,False,False,False
8,Silver,1,True,False,False,False,False,False,False,False,False,True,False,False,False,False,False,False
9,Silver,1,False,False,False,True,False,False,False,False,False,False,False,False,True,False,False,False


In [45]:
encoded_df.dtypes

customer_loyalty_tier                 str
Mapped_Loyalty                      int64
restaurant_city_Chicago              bool
restaurant_city_Houston              bool
restaurant_city_Los Angeles          bool
restaurant_city_New York             bool
restaurant_city_Philadelphia         bool
restaurant_city_Phoenix              bool
restaurant_city_San Antonio          bool
restaurant_city_San Diego            bool
restaurant_cuisine_American          bool
restaurant_cuisine_Chinese           bool
restaurant_cuisine_Fast Food         bool
restaurant_cuisine_Indian            bool
restaurant_cuisine_Italian           bool
restaurant_cuisine_Japanese          bool
restaurant_cuisine_Mediterranean     bool
restaurant_cuisine_Mexican           bool
dtype: object

### III. Working with other string data

In [62]:
#Identify a column containing unstructured text, mixed formats, or string artifacts (e.g., "$1,000", " Category A ", "ID-4920").
string_data_df = df[['menu_item_name', 'discount_code']]
string_data_df.head(10)

,menu_item_name,discount_code
0,Threat Sushi,PROMO-BFUZ
1,Sound Dessert,NaN
2,Whatever Main Course,NaN
3,Black Main Course,NaN
4,Physical Pasta,NaN
5,Sell Beverage,NaN
6,Turn Burger,NaN
7,Ago Sushi,NaN
8,Building Burger,PROMO-XWIT
9,Worry Pizza,NaN


In [65]:
#Perform basic string operations using the .str accessor (e.g., lowercasing, stripping trailing/leading whitespaces, replacing specific characters).
string_data_df['menu_item_name'] = string_data_df['menu_item_name'].str.lower().str.replace(' ', '_').str.strip()
string_data_df.head(10)

,menu_item_name,discount_code
0,threat_sushi,PROMO-BFUZ
1,sound_dessert,NaN
2,whatever_main_course,NaN
3,black_main_course,NaN
4,physical_pasta,NaN
5,sell_beverage,NaN
6,turn_burger,NaN
7,ago_sushi,NaN
8,building_burger,PROMO-XWIT
9,worry_pizza,NaN


In [71]:
nonull_string_data_df = df[['menu_item_name', 'discount_code']]
nonull_string_data_df = nonull_string_data_df.dropna(axis=0, how='any')
nonull_string_data_df.head(10)

,menu_item_name,discount_code
0,Threat Sushi,PROMO-BFUZ
8,Building Burger,PROMO-XWIT
12,Be Pizza,PROMO-GCOK
21,Different Beverage,PROMO-BZXO
24,Expect Pasta,PROMO-CXEK
26,According Pizza,PROMO-GPTV
31,Authority Sushi,PROMO-JZOF
34,Pretty Sushi,PROMO-GJKN
35,Person Burger,PROMO-FLPF
36,Rest Pasta,PROMO-JZNE


In [91]:
#Extract a specific pattern using Regular Expressions (Regex) via pandas string methods and save the extracted data as a new feature column.
nonull_string_data_df['promo_id'] = nonull_string_data_df['discount_code'].str.extract(r'PROMO-(.*)')
nonull_string_data_df.head(10)

,menu_item_name,discount_code,promo_id
0,Threat Sushi,PROMO-BFUZ,BFUZ
8,Building Burger,PROMO-XWIT,XWIT
12,Be Pizza,PROMO-GCOK,GCOK
21,Different Beverage,PROMO-BZXO,BZXO
24,Expect Pasta,PROMO-CXEK,CXEK
26,According Pizza,PROMO-GPTV,GPTV
31,Authority Sushi,PROMO-JZOF,JZOF
34,Pretty Sushi,PROMO-GJKN,GJKN
35,Person Burger,PROMO-FLPF,FLPF
36,Rest Pasta,PROMO-JZNE,JZNE


### IV. Balanced encoding

In [96]:
#Identify a categorical column with high cardinality (many unique categories), show cardinality statistics by categorical columns.

cols = df.select_dtypes(include=['str']).columns
for col in cols:
    print(f"{col}: {df[col].nunique()} unique values")

customer_loyalty_tier: 3 unique values
customer_acquisition_channel: 5 unique values
customer_signup_date: 730 unique values
restaurant_cuisine: 8 unique values
restaurant_city: 8 unique values
menu_item_name: 1034 unique values
menu_item_category: 8 unique values
order_timestamp: 10000 unique values
delivery_timestamp: 9224 unique values
discount_code: 1949 unique values
discount_type: 3 unique values
order_status: 2 unique values
event_flag: 1 unique values


In [123]:
#menu_item_name column has a high cardinality
#Apply a balanced/advanced encoding technique (such as Target Encoding, Frequency/Count Encoding, or Hashing) to this column.

#Frequency Encoding
encoding_df = df[['menu_item_name', 'menu_item_id', 'total_amount']]
freq_map = encoding_df['menu_item_name'].value_counts()
encoding_df['menu_item_freq'] = encoding_df['menu_item_name'].map(freq_map)
encoding_df.head(10)

,menu_item_name,menu_item_id,total_amount,menu_item_freq
0,Threat Sushi,669,25.42,12
1,Sound Dessert,231,22.70,10
2,Whatever Main Course,1039,27.74,8
3,Black Main Course,487,31.85,5
4,Physical Pasta,442,28.46,7
5,Sell Beverage,878,64.00,26
6,Turn Burger,274,139.07,7
7,Ago Sushi,618,37.19,15
8,Building Burger,1027,38.93,9
9,Worry Pizza,326,18.76,11


In [124]:
#Target Encoding
target_mean = encoding_df.groupby('menu_item_name')['total_amount'].mean()

encoding_df['menu_item_target_enc'] = encoding_df['menu_item_name'].map(target_mean)

encoding_df.head(10)

,menu_item_name,menu_item_id,total_amount,menu_item_freq,menu_item_target_enc
0,Threat Sushi,669,25.42,12,36.559167
1,Sound Dessert,231,22.70,10,21.914000
2,Whatever Main Course,1039,27.74,8,32.862500
3,Black Main Course,487,31.85,5,31.786000
4,Physical Pasta,442,28.46,7,36.991429
5,Sell Beverage,878,64.00,26,47.786538
6,Turn Burger,274,139.07,7,57.050000
7,Ago Sushi,618,37.19,15,37.387333
8,Building Burger,1027,38.93,9,39.872222
9,Worry Pizza,326,18.76,11,25.919091


#### Observe how this affects the dataset compared to standard One-Hot Encoding.
- Original Columns: 1

- One-Hot Encoding would create: 1034 columns (one column for each unique)

- Frequency Encoding created: 1 column

- Target Encoding created: 1 column
